src/vision/status_quo/alan_writes_a_web_framework.md

```diff
@@ -0,0 +1,222 @@
+
+fn get_products_handler(state: State) -> Pin<Box<HandlerFuture>> {
+    use crate::schema::products::dsl::*;
+
+    async move {
+        let repo = Repo::borrow_from(&state);
+        let result = repo.run(move |conn| products.load::<Product>(&conn)).await;
+        match result {
+            Ok(prods) => {
+                let body = serde_json::to_string(&prods).expect("Failed to serialize prods.");
+                let res = create_response(&state, StatusCode::OK, mime::APPLICATION_JSON, body);
+                Ok((state, res))
+            }
+            Err(e) => Err((state, e.into())),
+        }
+    }
+    .boxed()
+}
+```
+and then it is registered like this:
+```rust
+    router_builder.get("/").to(get_products_handler);
+```
+
+The handler code is forced to drift to the right a lot, because of the async block, and the lack of ability to use `?` forces the use of a match block, which drifts even further to the right. This goes against [what he has learned from his days writing go](https://github.com/uber-go/guide/blob/master/style.md#reduce-nesting).
+
+Rather than switching YouBuy to a different web framework, Alan decides to contribute to the web framework himself. After a bit of a slog and a bit of where-clause-soup, he manages to make the web framework capable of using an `async fn` as an http request handler. He does this by extending the router builder with a closure that boxes up the `impl Future` from the async fn and then passes that closure on to `.to()`.
+
+```rust
+    fn to_async<H, Fut>(self, handler: H)
+    where
+        Self: Sized,
+        H: (FnOnce(State) -> Fut) + RefUnwindSafe + Copy + Send + Sync + 'static,
+        Fut: Future<Output = HandlerResult> + Send + 'static,
+    {
+        self.to(move |s: State| handler(s).boxed())
+    }
+```
+The handler registration then becomes:
+```rust
+    router_builder.get("/").to_async(get_products_handler);
+```
+
+This allows him to strip out the async blocks in his handlers and use `async fn` instead.
+
+```rust
+// Type the library again, in case you've forgotten:
+pub type HandlerResult = Result<(State, Response<Body>), (State, HandlerError)>;
+
+async fn get_products_handler(state: State) -> HandlerResult {
+    use crate::schema::products::dsl::*;
+
+    let repo = Repo::borrow_from(&state);
+    let result = repo.run(move |conn| products.load::<Product>(&conn)).await;
+    match result {
+        Ok(prods) => {
+            let body = serde_json::to_string(&prods).expect("Failed to serialize prods.");
+            let res = create_response(&state, StatusCode::OK, mime::APPLICATION_JSON, body);
+            Ok((state, res))
+        }
+        Err(e) => Err((state, e.into())),
+    }
+}
+```
+
+It's still not fantastically ergonomic though. Because the handler takes ownership of State and returns it in tuples in the result, Alan can't use the `?` operator inside his http request handlers. If he tries use `?` in a handler, like this:
```

Comment:
```suggestion
It's still not fantastically ergonomic though. Because the handler takes ownership of State and returns it in tuples in the result, Alan can't use the `?` operator inside his http request handlers. If he tries to use `?` in a handler, like this:
```